File: ios/NotificationServiceExtension/NotificationService.m
//  SPDX-License-Identifier: ice License 1.0
//
//  NotificationService.m
//  NotificationServiceExtension
//

#import "NotificationService.h"
#import "FirebaseMessaging.h"

@interface NotificationService ()

@property (nonatomic, strong) void (^contentHandler)(UNNotificationContent *contentToDeliver);
@property (nonatomic, strong) UNMutableNotificationContent *bestAttemptContent;

@end

@implementation NotificationService

- (void)didReceiveNotificationRequest:(UNNotificationRequest *)request withContentHandler:(void (^)(UNNotificationContent * _Nonnull))contentHandler {
    self.contentHandler = contentHandler;
    self.bestAttemptContent = [request.content mutableCopy];
    
    NSString *appLocale = [self getAppLocale];

    if (appLocale != nil) {
        // Get translations
        // Match translations with the data in the notification
        self.bestAttemptContent.body = appLocale;
    }
    
    [[FIRMessaging extensionHelper] populateNotificationContent:self.bestAttemptContent withContentHandler:contentHandler];
}

- (void)serviceExtensionTimeWillExpire {
    self.contentHandler(self.bestAttemptContent);
}

- (NSString *)getAppLocale {
    NSString *appGroup = [[NSBundle mainBundle] objectForInfoDictionaryKey:@"APP_GROUP"];
    if (appGroup) {
        NSUserDefaults *userDefaults = [[NSUserDefaults alloc] initWithSuiteName:appGroup];
        NSString *appLocale = [userDefaults objectForKey:@"app_locale"];
        return appLocale;
    }
    return nil;
}

@end

File: ios/NotificationServiceExtension/NotificationService.swift
// SPDX-License-Identifier: ice License 1.0

import FirebaseMessaging
import UserNotifications

class NotificationService: UNNotificationServiceExtension {
    var contentHandler: ((UNNotificationContent) -> Void)?
    var mutableNotificationContent: UNMutableNotificationContent?

    override func didReceive(
        _ request: UNNotificationRequest,
        withContentHandler contentHandler: @escaping (UNNotificationContent) -> Void
    ) {
        self.contentHandler = contentHandler
        mutableNotificationContent = (request.content.mutableCopy() as? UNMutableNotificationContent)

        guard let mutableNotificationContent = mutableNotificationContent else {
            contentHandler(request.content)
            return
        }

        Task {
            do {
                let translation = await NotificationTranslationService(storage: try SharedStorageService()).translate(request.content.userInfo)

                if let title = translation.title, let body = translation.body {
                    mutableNotificationContent.title = title
                    mutableNotificationContent.body = body
                }
            } catch {
                NSLog("Failed to translate notification: \(error)")
            }

            Messaging.serviceExtension().populateNotificationContent(
                mutableNotificationContent,
                withContentHandler: contentHandler
            )
        }
    }

    override func serviceExtensionTimeWillExpire() {
        if let contentHandler = contentHandler, let mutableNotificationContent = mutableNotificationContent {
            contentHandler(mutableNotificationContent)
        }
    }
}
